### PR TITLE
fix(prefer-in-document): false positive on .toHaveLength(1) matcher with *AllBy* query

### LINKED ISSUE #171: prefer-in-document: `.toHaveLength(1)` vs `.toBeInTheDocument()`
- `eslint-plugin-jest-dom` version: 3.8.1
- `node` version: 16.1.0
- `npm` version: 7.11.1

Relevant code or config

```javascript
expect(screen.queryAllByRole('option')).toHaveLength(2);
// some action here
expect(screen.queryAllByRole('option')).toHaveLength(1);
```

What you did:
I want to ensure there is only one `option`.

What happened:

The rule wants this to be
```js
expect(screen.queryByRole('option')).toBeInTheDocument();
```

Reproduction repository:

Problem description:
It's not the same thing! I cannot assert that there is only one `option` now.

Suggested solution:
`prefer-in-document` should allow `.toHaveLength(1)`. Or there should be an option at least.

### PR DESCRIPTION
**Pull request by SevenOutman**
**What**:

Close #171 

**Why**:

It's valid to use `.toHaveLength(1)` with `*AllBy*` queries to check for "exactly one match".

See conclusion of discussions at https://github.com/testing-library/eslint-plugin-jest-dom/issues/171#issuecomment-895074086

**How**:

Implement the truth table as described in the link above.

**Checklist**:

- [x] Documentation
- [x] Tests
- [x] Ready to be merged

### CODE REVIEW

**Comment by G-Rath** [comment]
hey sorry for the delay - would you mind rebasing this so that we've got the CI checks going? then I'll have a review

**Comment by SevenOutman** [comment]
@G-Rath Sure, I have rebased it.

File: src/rules/prefer-in-document.js [hunk 1]
```diff
@@ -43,6 +45,24 @@ function usesToHaveLengthZero(matcherNode, matcherArguments) {
...
 }
 
+/**
+ * Extract the DTL query identifier from a call expression
+ *
+ * <query>() -> <query>
+ * screen.<query>() -> <query>
+ */
+function getDTLQueryIdentifierNode(callExpressionNode) {
+  if (!callExpressionNode || callExpressionNode.type !== "CallExpression") {
```
Comment (G-Rath) [inline src/rules/prefer-in-document.js:55]: I'm pretty sure we can use optional chaining here:

```suggestion
  if (callExpressionNode?.type !== "CallExpression") {
```

File: src/rules/prefer-in-document.js [hunk 2]
```diff
@@ -78,21 +98,64 @@ export const create = (context) => {
...
-        isNotToHaveLengthZero ||
-        !["Literal", "Identifier"].includes(matcherArguments[0].type) ||
-        lengthValue === undefined ||
-        lengthValue > 1;
-
-      if (isValidUseOfToHaveLength) {
+      const queryName = queryNode.name || queryNode.property.name;
+
+      const isSingleQuery =
+        queries.includes(queryName) && !/AllBy/.test(queryName);
```
Comment (G-Rath) [inline src/rules/prefer-in-document.js:115]:
```suggestion
        queries.includes(queryName) && !/AllBy/u.test(queryName);
```

File: src/rules/prefer-in-document.js [hunk 3]
```diff
@@ -78,21 +98,64 @@ export const create = (context) => {
...
+          node: matcherNode,
+          messageId: "invalid-combination-length-1",
+          data: {
+            query: queryName,
+            allQuery,
+          },
+          loc: matcherNode.loc,
+          suggest: [
+            {
+              desc: `Replace ${queryName} with ${allQuery}`,
```
Comment (G-Rath) [inline src/rules/prefer-in-document.js:136]: Let's use `messageId` for these instead

File: src/rules/prefer-in-document.js [hunk 4]
```diff
@@ -184,6 +247,12 @@ export const create = (context) => {
         context,
         node.object.object.arguments[0].name
       );
+
+      // Not an RTL query
+      if (!queryNode || queryNode.type !== "CallExpression") {
```
Comment (G-Rath) [inline src/rules/prefer-in-document.js:252]: I think we can use optional chaining here:

```suggestion
      if (queryNode?.type !== "CallExpression") {
```

File: src/rules/prefer-in-document.js [hunk 5]
```diff
@@ -201,16 +270,25 @@ export const create = (context) => {
...
     ) {
-      const queryNode = getAssignmentForIdentifier(
+      // Value expression being assigned to the left-hand value
+      const rightValueNode = getAssignmentForIdentifier(
         context,
         node.object.arguments[0].name
       );
+
+      // Not a DTL query
+      if (!rightValueNode || rightValueNode.type !== "CallExpression") {
```
Comment (G-Rath) [inline src/rules/prefer-in-document.js:280]: I think we can use optional chaining here:

```suggestion
      if (rightValueNode?.type !== "CallExpression") {
```

**Review by G-Rath — CHANGES_REQUESTED** [review]
Looks good - got some minor changes I'd like made, and you also need to rebase again (sorry 😅) and get CI passing (you're missing a bit of coverage)

File: src/rules/prefer-in-document.js (hunk 1 above)
Comment (SevenOutman) [reply src/rules/prefer-in-document.js:55]: I applied the suggestion but it turned out that ESLint does not recognize optional chaining syntax in this project.

File: src/rules/prefer-in-document.js (hunk 3 above)
Comment (SevenOutman) [reply src/rules/prefer-in-document.js]: Replaced with `messageId`

**Comment by SevenOutman** [comment]
All right, I've merged `main` branch and updated the suggestion message.

**Comment by G-Rath** [comment]
@SevenOutman if you're still interested in this, could you resolve the test failures?

**Comment by SevenOutman** [comment]
> @SevenOutman if you're still interested in this, could you resolve the test failures?

Sorry, I'm afraid I don't have enough mind bandwidth for this edit right now.

**Comment by EricKim987** [comment]
Can I try taking over this pr? Or should I just create a new one? I really want to fix this. @G-Rath

**Comment by G-Rath** [comment]
@EricKim987 go for it - I think it would be best to create a new PR though

**Pull request closed without merging**